### ISSUE TITLE
F3 (show find dialog) is handled 2 times

### ISSUE THREAD

**Issue by OmlineEditor** [Alexey-T/CudaLister#113]
Я нажимаю в листере Ctrl+F но окно поиска не появляется, вы его перенести на F3 но и при нажитии на F3 тоже не всегда появляется это окно, через раз как-то появляется. а когда появилось я его не могу закрыть с первого раза, я его закрываю по Esc или на крестик давлю, а оно не запрывается с первого раза, только со второго раза. Неудобно стало :-(

Прошу починить поиск и сделать чтобы окно вседа появлялось по Ctrl+F, а при нажатии Esc закрывалось с первого раза.

Windows 7 x64
TC 10.50
cudalister 1.8.1

**Comment by Alexey-T** [comment]
это я смогу поделать только через неделю-две, когда буду под виндой.

**Comment by OmlineEditor** [comment]
Хорошо, я подожду помучаюсь пока. Вы только исправьте баг.

**Comment by Alexey-T** [comment]
TotalCmd у вас 32-битный? это бы обьяснило почему ctrl+f не работает. оно запрещено в 32 бит версии. чтобы избежать бага вставки в текст 'F'

**Comment by OmlineEditor** [comment]
> TotalCmd у вас 32-битный? это бы обьяснило почему ctrl+f не работает. оно запрещено в 32 бит версии. чтобы избежать бага вставки в текст буквы 'F'

Это потому что в плагине еще можно редактировать текст? В других программах нормально работает поиск и CTRL+F и ничего не вставляется лишнего. Может эту проблему как-то можно решить? Это наверное в прошлом было и сейчас явно должно быть решение

**Comment by Alexey-T** [comment]
это проблема с первой версии когда появилось редактирование. решения нет.

**Comment by OmlineEditor** [comment]
Я думаю решение есть но оно очень сложное и запутанное. Ладно буду привыкать. Вы тогла хоть почините закрывание окна поиска по Esc и крестику с первого раза.

**Comment by OmlineEditor** [comment]
полгода прошло, как там с решением? может в режиме толко чтения сделать нормальный поисх через ctrl+F ?

**Comment by Alexey-T** [comment]
попробовал поправить, но не получается. в х32 версии Тотала. почему-то приходит два нажатия F3 вместо одного вот и окно F3 появляется после Esc. может кто-то с форума Тотала сможет поправить.

**Comment by OmlineEditor** [comment]
Покажите полный кусок кода который отвечает за работу Ctrl+F, попробую посмотреть, может докумекаю где ошибка.

**Comment by Alexey-T** [comment]
form_main.pas
```

  //support Ctrl+F
  if (Key=VK_F) and (Shift=[ssCtrl]) then
  begin
    //OptRep:= true;
    DoFindDialog;
    Key:= 0;
    exit
  end;
```

выше этого куска идет проверка на дублированное нажатие (особеность листера)
```
  //to ignore duplicate keys (some Lazarus bug)
  MaybeDups:=
```

**Comment by OmlineEditor** [comment]
А можно чтобы предотвратить появление в тексте буквы F в это условие добавить что поиск будет только в режиме чтения? так можно будет сделать для работы, чтобы заработала клавиша Ctrl+F?
`if (Key=VK_F) and (Shift=[ssCtrl]) then`

В этом коде вроде все хорошо у вас. Можно теперь показать код с обработкой F3, я попробую понять почему задваивается клавиша и не появляется окно поиска?

**Comment by Alexey-T** [comment]
да, это оно. обработка VK_F3.

>в это условие добавить что поиск будет только в режиме чтения? 

попробуйте. readonly это `ed.ModeReadOnly`.

**Comment by OmlineEditor** [comment]
По поводу двух нажатий клавиши F3 возможно это из-за того что в обработке нажатия F3 код `Key:= 0;` может не выполнится при условии что выполнится этот блок и не дойдет до обнуления Key.
```
 if Finder.StrFind='' then
    begin
      DoFindDialog;
      exit;
    end;
```

Может тогда вот так сделать?
```  
  //support F3, find next
  if (Key=VK_F3) and (Shift=[]) then
  begin
    Key := 0; // Обнуляем значение Key, чтобы предотвратить дальнейшую обработку клавиши
    //DoFind(true, true, Finder.OptCase, Finder.OptWords, '');
    Finder.StrFind:= GetLastFindText;
    if ed.TextSelected<>'' then begin
      Finder.StrFind:= ed.TextSelected;
      SetLastFindText(Finder.StrFind);
    end;
    if Finder.StrFind='' then
    begin
      DoFindDialog;
      exit;
    end;
    Finder.OptBack:= false;
    Finder.OptFromCaret:= true;
    ok:= Finder.DoAction_FindOrReplace(false, false, bChanged, true);
    FinderUpdateEditor(false);
    if not ok then
    begin
      DoFindError;
      DoFindDialog;
    end;
//    Key:= 0;  // это мы переносим выше
    exit
  end;
```


Еще по поводу работы Esc, где там закрытие окна поиска? Я в коде не вижузакрытие окна, мои познания слабые, может что-то проглядела? Я раньше писала что не могу его закрыть по Esc.

**Comment by Alexey-T** [comment]
Вы перенесли обнуление Key выше. но в блоке Key вроде бы не использовался, тогда без разницы - где мы обнуляем Key.

**Comment by OmlineEditor** [comment]
> Вы перенесли обнуление Key выше. но в блоке Key вроде бы не использовался, тогда без разницы - где мы обнуляем Key.

вот в том-то и проблема что если у нас выплнилось условие `if Finder.StrFind='' then` мы не обнулим key и у нас может быть двойное нажатие. Или не будет и я что-то путаю?

**Comment by Alexey-T** [comment]
Вы просто даете идеи или сами правите и сами собираете + тестите? Надо тестить вам еще. А то я под линукс и ваши идеи даже проверить не могу тут

**Comment by Alexey-T** [comment]
>Компилятор ставится просто или с танцами и с бубном?

Трудно. так что не надо тогда.

**Comment by OmlineEditor** [comment]
Да, щас закрывается по Esc с первого раза, спасибо за решение глюка.
Кстати вернуть назад поиск по CTRL+F все еще не вариант?

**Comment by Alexey-T** [comment]
Ситуация с Ctrl+F пока не изменилась.
